### PR TITLE
Demystify project - first PR

### PR DESCRIPTION
**Pull request by JonathanYK**
Demystify project - first PR

### CODE REVIEW

File: src/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (arseni-meteor) [inline src/.vscode/launch.json:1]: .vscode (and even usually .* folders) should not be git-tted

Comment (arseni-meteor) [inline src/.vscode/launch.json:1]: .vscode (and even usually .* folders) should not be git-ted

File: src/bp/home_bp.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from project.sessions import db, Sessions
+
+# validates that the provided json file includes mandatory fields: 
+def validate_json_keys(json_data):
+    keys_dict = json_data.keys()
+    must_have_keys = ["message", "additional", "level", "timestamp", "fileName", "lineNumber"]
+
+    return all(map(lambda provided_json_keys: True if provided_json_keys in keys_dict else False, must_have_keys))
+
+# sets the provided msg in the db, handles overriding data:
```
Comment (arseni-meteor) [inline src/bp/home_bp.py:22]: Great comment

File: src/bp/home_bp.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        ret_val = ""
+        session_update = db.session.query(Sessions).get(session_id)
+        if session_update.msg :
+            ret_val = "Overriding message for session id: " + session_id + ", previous message was: " + session_update.msg + "\n"
+        session_update.msg = msg
+        db.session.commit()
+        return ret_val
+
+    except:
+        return handle_server_500("There was an issue adding message to a session id!")
```
Comment (arseni-meteor) [inline src/bp/home_bp.py]: This explains what happened, which is nice
But when a person, or, the responsible developer (YOU) will here from a customer that he has a problem, and after you ask what he see's, he will tell you that "There was an issue adding message to a session id!" - What will you do? Pretty stuck right there..
Add the session id and msg in the error to have an edge to start searching from (which you will do when this error is encountered)

File: src/bp/log_bp.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import logging
+import os
+from flask import Blueprint, request
+from .err_bp import handle_bad_input_501
+from .home_bp import set_msg, validate_json_keys
+from .sessions_bp import session_id_validation
+
+def log_level_validation(cur_log_level):
+    curr_log_lvls = ["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG", "NOTSET"]
+    return True if cur_log_level in curr_log_lvls else False
```
Comment (arseni-meteor) [inline src/bp/log_bp.py]: nnnice

File: src/bp/log_bp.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+def print_DEBUG(logger, logging_data):
+    for data in logging_data: logger.debug(data)
+
+def print_NOTSET(logger, logging_data):
+    for data in logging_data: logger.notset(data)
+
+
+# logging the data to a .log file:
+def log_data(logger, session_id, json_data):
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:38]: OOP- maybe keep the logger as a member in some class?

File: src/bp/log_bp.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    # list for relevant logging params:
+    logging_data = []
+
+    # parsering the json params:
+    str_log_level = json_data["level"]
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:44]: If the json_data doesn't have a "level" field?
Error handeling..

File: src/bp/log_bp.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    if log_level_validation(str_log_level) is False:
+        return handle_bad_input_501("The provided log level isn't valid!")
+
+    logging_data.append(str_log_level)
+    logging_data.append(json_data["timestamp"])
+    logging_data.append(json_data["fileName"])
+    logging_data.append(json_data["lineNumber"])
+
+    # create log level printings according to provided log level, then set the same log level:
+    dynamic_log_printer = globals()["print_%s" % str_log_level]
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:56]: Cool

File: src/bp/log_bp.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    logger.setLevel(logging.getLevelName(str_log_level))
+
+    # config session formatter both for file_handler and stream_handler:
+    session_formatter = logging.Formatter("%(asctime)s::%(levelname)s::%(name)s::%(message)s")
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:60]: Very cool

File: src/bp/log_bp.py (hunk 5 above)
Comment (arseni-meteor) [inline src/bp/log_bp.py:38]: Add a unit test (+error oriented) for this function. Heavily used

File: src/bp/log_bp.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    file_handler = logging.FileHandler(filename=log_filename, mode="a")
+    file_handler.set_name(session_id)
+    file_handler.setFormatter(session_formatter)
+
+    # stream_handler configurations:
+    stream_handler = logging.StreamHandler()
+    stream_handler.setFormatter(session_formatter)
+
+    logger.addHandler(file_handler)
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:74]: Amm... why is this needed?

File: src/bp/log_bp.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    # removing both file and stream handlers, in order to avoid duplications:
+    logger.removeHandler(file_handler)
+    logger.removeHandler(stream_handler)
+
+
+log_blueprint = Blueprint('log', __name__)
+
+# handles invalid requests:
+@log_blueprint.route("/log", methods=["GET", "PUT", "DELETE"])
+def handle_session_log_forbidden_requests():
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:89]: Check if theres a default solution for that in Flask

File: src/bp/log_bp.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    return handle_bad_input_501("This method is forbidden for log url!")
+
+# handles posting log to db:
+@log_blueprint.route("/log", methods=["POST"])
+def log():
+
+    ret_val = ""
+
+    # retriving the session id from the url:
+    curr_session_id = request.args.get("session_id")
```
Comment (arseni-meteor) [inline src/bp/log_bp.py:99]: If it's not there..?

File: src/bp/log_bp.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    # validate the session id:
+    if session_id_validation(curr_session_id) is False:
+        return handle_bad_input_501("The provided session id isn't valid!")
+
+    # validate there arn't missing required keys in json file:
+    if validate_json_keys(request.json) is False:
+        return handle_bad_input_501("There is a missing key/s in the provided json!")
```
Comment (arseni-meteor) [inline src/bp/log_bp.py]: Which one..?
Add all the existing, maybe the diff with what you expect
This is high level^

File: src/bp/sessions_bp.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+def session_id_validation(curr_session_id):
+    return db.session.query(db.exists().where(Sessions.id == curr_session_id)).scalar()
+
+def get_new_session_id():
+    
+    # generate random int with exact 9 integers to be a session_id:
+    gen_rand_session = random.randrange(100000000, 999999999, 9)
+
+    # validate that gen_rand_session not exists in the db:
+    while db.session.query(db.exists().where(Sessions.id == gen_rand_session)).scalar():
```
Comment (arseni-meteor) [inline src/bp/sessions_bp.py:17]: Nice

File: src/bp/sessions_bp.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+         gen_rand_session = random.randrange(100000, 999000, 6)
+
```
Comment (arseni-meteor) [inline src/bp/sessions_bp.py:19]: You can leave just the while somehow (do while would be classic)
It's ok like that anyway

File: src/bp/sessions_bp.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    
+    try:
+        ret_msg = db.session.query(Sessions).get(session_id).msg
+        if not ret_msg:
+            return "There is no message for session id " + session_id
+        else :
+            return "The message for session id " + str(session_id) + " is: " + ret_msg
+    
+    except:
+         return handle_server_500("There was an issue retriving the message for this session id!")
```
Comment (arseni-meteor) [inline src/bp/sessions_bp.py]: Which.... one!

File: src/bp/sessions_bp.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+
+sessions_blueprint = Blueprint('sessions', __name__)
+
+# handles invalid requests:
+@sessions_blueprint.route("/session_id", methods=["POST", "PUT", "DELETE"])
+def handle_session_id_forbidden_requests():
+    return handle_bad_input_501("This method is forbidden for session_id url!")
+
+# TODO: add mutex for accessing db, then add test with 10,000 requests:
```
Comment (arseni-meteor) [inline src/bp/sessions_bp.py]: Yas! This will be great

File: src/bp/sessions_bp.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+# retriving the message of a provided session_id:
+@sessions_blueprint.route("/session_msg", methods=["GET"])
+def get_session_msg():
+    
+    # retriving the session_id from the request:
+    curr_session_id = request.args.get("session_id")
+
+    if session_id_validation(curr_session_id) is False:
+        return handle_bad_input_501("The provided session id isn't valid!")
```
Comment (arseni-meteor) [inline src/bp/sessions_bp.py]: Ra

File: src/instance/flask.cfg [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+##########################################################
+import os
+
+
+# Get the folder of the top-level directory of this project
+BASEDIR = os.path.abspath(os.path.dirname(__file__))
+
+# Update later by using a random number generator and moving
+# the actual key outside of the source code under version control
+SECRET_KEY = 'bad_secret_key'
```
Comment (arseni-meteor) [inline src/instance/flask.cfg]: Yes.. indeed bad one
Generate something new

File: src/instance/flask_test.cfg [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+##########################################################
+import os
+
+
+# Get the folder of the top-level directory of this project
+BASEDIR = os.path.abspath(os.path.dirname(__file__))
+
+# Update later by using a random number generator and moving
+# the actual key outside of the source code under version control
+SECRET_KEY = 'bad_secret_key'
```
Comment (arseni-meteor) [inline src/instance/flask_test.cfg]: SSOT

File: src/project/__init__.py [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    app.config.from_pyfile(config_filename)
+    return app
+
+def db_config(app, db_name):
+    
+    app.config['SQLALCHEMY_DATABASE_URI'] = f'sqlite:///{db_name}'
+    db.init_app(app)
+    
+    # in case the're is no db_name database:
+    if not database_exists(f'sqlite:///project/{db_name}'):
```
Comment (arseni-meteor) [inline src/project/__init__.py]: Maybe this if should be before? (Im not sure)

File: src/project/sessions.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
+# sessions table:
```
Comment (arseni-meteor) [inline src/project/sessions.py]: Maybe remove the 'sessions.db' file

File: src/tests/functionality/test_sessions.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
+
+def test_main_page(test_client):
+    """
+    GIVEN a Flask application configured for testing
```
Comment (arseni-meteor) [inline src/tests/functionality/test_sessions.py:3]: Very very nice!

File: src/tests/functionality/test_sessions.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    response = test_client.get('/session_id')
+    ret_lst = str(response.data)[2:-1].split(":")
+    gen_session_id = int(ret_lst[1][1:])
+
+    assert ret_lst[0] == "Session ID generated"
+    assert gen_session_id >= 1000
+    assert response.status_code == 201
+
+    if return_session_id:
+        return gen_session_id
```
Comment (arseni-meteor) [inline src/tests/functionality/test_sessions.py]: Fail oriented ones

File: src/tests/unit/test_unit.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+    # sample message to test:
+    json_msg_for_testing = "This is a test message from test_sessions_id_msg"
+
+    # send sample json, and retrive it's session_id:
+    gen_session_id = test_send_json(test_client=test_client, json_msg_for_testing=json_msg_for_testing)
+
+    # get response message of gen_session_id:
+    response = test_client.get(f'/session_msg?session_id={gen_session_id}')
+
+    assert f"{json_msg_for_testing} session_id: " + str(gen_session_id) in str(response.data)
```
Comment (arseni-meteor) [inline src/tests/unit/test_unit.py]: Add fail oriented (catch exception/assert_exception)

**Review by arseni-meteor — CHANGES_REQUESTED** [review]
Very very nice all in all, some slight fixes

File: src/.vscode/launch.json (hunk 1 above)
Comment (JonathanYK) [reply src/.vscode/launch.json:1]: right, forgot to add it to.gitignore :\

File: src/bp/home_bp.py (hunk 2 above)
Comment (JonathanYK) [reply src/bp/home_bp.py:22]: 👍

File: src/.vscode/launch.json (hunk 1 above)
Comment (arseni-meteor) [reply src/.vscode/launch.json:1]: Have you went over the PR before adding me as a reviewer?

Comment (JonathanYK) [reply src/.vscode/launch.json:1]: Yes, first I thought that file should be included, now checked and understand that should be ignored!

File: src/bp/home_bp.py (hunk 3 above)
Comment (JonathanYK) [reply src/bp/home_bp.py]: 100% agree, changed.

File: src/bp/log_bp.py (hunk 4 above)
Comment (JonathanYK) [reply src/bp/log_bp.py]: 👍

File: src/bp/log_bp.py (hunk 5 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:38]: Can change it, but is that needed? this is the only method in this class... If you think it is better, I'll customize it

File: src/bp/log_bp.py (hunk 6 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:44]: There is a method 'validate_json_keys' on home_bp.py that handles this.

File: src/bp/log_bp.py (hunk 7 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:56]: 👍

File: src/bp/log_bp.py (hunk 8 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:60]: 👍 👍

File: src/bp/log_bp.py (hunk 5 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:38]: There is a test for this - I'll add an error oriented (missing)

File: src/bp/log_bp.py (hunk 9 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:74]: The correct way in python logging is to use handlers (both stream and file - for printing the log to the console/to a file, accordingly).

File: src/bp/log_bp.py (hunk 10 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:89]: Searched again, did not found anything else (I used this approach after I saw it in a few Flask examples..)

File: src/bp/log_bp.py (hunk 11 above)
Comment (JonathanYK) [reply src/bp/log_bp.py:99]: in the same file row 99 "if session_id_validation(curr_session_id) is False:"
So if there is no session_id in the route, this if will be true :)

File: src/bp/sessions_bp.py (hunk 13 above)
Comment (JonathanYK) [reply src/bp/sessions_bp.py:17]: 👍

File: src/bp/sessions_bp.py (hunk 14 above)
Comment (JonathanYK) [reply src/bp/sessions_bp.py:19]: 👍

File: src/bp/sessions_bp.py (hunk 15 above)
Comment (JonathanYK) [reply src/bp/sessions_bp.py]: Added 👍

File: src/bp/sessions_bp.py (hunk 17 above)
Comment (JonathanYK) [reply src/bp/sessions_bp.py]: Added 👍

File: src/project/__init__.py (hunk 20 above)
Comment (JonathanYK) [reply src/project/__init__.py]: "AssertionError: The sqlalchemy extension was not registered to the current application.  Please make sure to call init_app() first" 
I have no choice :\

File: src/project/sessions.py (hunk 21 above)
Comment (JonathanYK) [reply src/project/sessions.py]: right, not needed here - will be removed.

File: src/tests/functionality/test_sessions.py (hunk 22 above)
Comment (JonathanYK) [reply src/tests/functionality/test_sessions.py:3]: 👍 👍 👍

File: src/bp/log_bp.py (hunk 12 above)
Comment (JonathanYK) [reply src/bp/log_bp.py]: Thanks, added!

File: src/instance/flask.cfg (hunk 18 above)
Comment (JonathanYK) [reply src/instance/flask.cfg]: Done - reading secret key from a dedicated file

File: src/instance/flask_test.cfg (hunk 19 above)
Comment (JonathanYK) [reply src/instance/flask_test.cfg]: Done - reading secret key from a dedicated file

File: src/bp/sessions_bp.py (hunk 16 above)
Comment (JonathanYK) [reply src/bp/sessions_bp.py]: Done on latest commit

File: src/tests/functionality/test_sessions.py (hunk 23 above)
Comment (JonathanYK) [reply src/tests/functionality/test_sessions.py]: Done on the latest commit

File: src/tests/unit/test_unit.py (hunk 24 above)
Comment (JonathanYK) [reply src/tests/unit/test_unit.py]: Done on the latest commit

**Comment by JonathanYK** [comment]
@seniakalma @arseni-meteor 
Please review the whole project for the last time after my 2 latest commits. 
Last stage- readme (WIP)

File: src/.vscode/launch.json (hunk 1 above)
Comment (seniakalma) [reply src/.vscode/launch.json:1]: Hope it's gone

File: src/bp/log_bp.py (hunk 5 above)
Comment (seniakalma) [reply src/bp/log_bp.py:38]: We talked about this

File: src/bp/log_bp.py (hunk 6 above)
Comment (seniakalma) [reply src/bp/log_bp.py:44]: We talked (put it in the beginning)

File: src/bp/log_bp.py (hunk 5 above)
Comment (seniakalma) [reply src/bp/log_bp.py:38]: Let's see the tests today

File: src/bp/log_bp.py (hunk 9 above)
Comment (seniakalma) [reply src/bp/log_bp.py:74]: V

File: src/bp/log_bp.py (hunk 10 above)
Comment (seniakalma) [reply src/bp/log_bp.py:89]: V

File: src/bp/log_bp.py (hunk 11 above)
Comment (seniakalma) [reply src/bp/log_bp.py:99]: V

File: src/bp/sessions_bp.py (hunk 14 above)
Comment (seniakalma) [reply src/bp/sessions_bp.py:19]: Found a bug there (different constants used)

File: src/bp/sessions_bp.py (hunk 16 above)
Comment (seniakalma) [reply src/bp/sessions_bp.py]: I want to hear all about the test!

File: src/instance/flask.cfg (hunk 18 above)
Comment (seniakalma) [reply src/instance/flask.cfg]: Secret (.name) file that isn't in the git

File: src/instance/flask_test.cfg (hunk 19 above)
Comment (seniakalma) [reply src/instance/flask_test.cfg]: V

File: src/project/__init__.py (hunk 20 above)
Comment (seniakalma) [reply src/project/__init__.py]: V

File: src/project/sessions.py (hunk 21 above)
Comment (seniakalma) [reply src/project/sessions.py]: V

File: src/tests/functionality/test_sessions.py (hunk 23 above)
Comment (seniakalma) [reply src/tests/functionality/test_sessions.py]: Let's see today

**Comment by seniakalma** [comment]
No new comments! Great work
Let's see the tests today

**Comment by JonathanYK** [comment]
> No new comments! Great work Let's see the tests today

Great! viewed all the additional comments.
I'll change the project&branch names and merge.

PROJECT DONE.
